astronomaly/feature_extraction/shape_features.py

```diff
@@ -293,107 +391,155 @@ def _execute_function(self, image):
-                prms = feats[n]
-                residuals.append(prms[0])
-                if prms[1] == 0 or prms[2] == 0:
-                    r = 0
+                    failed = True
+                    failure_message = "No contour found"
+
+                if failed:
+                    feats.append([np.nan] * 5)
+                    logging_tools.log(failure_message)
+
+                # Now we have the leastsq value, x0, y0, aspect_ratio, theta for each 
+                # sigma
+                # Normalise things relative to the highest threshold value
+                # If there were problems with any sigma levels, set all values to NaNs
+                if np.any(np.isnan(feats)):
+                    return [np.nan] * 4 * len(self.percentiles)
                 else:
-                    x_diff = prms[1] - x0_max_sigma
-                    y_diff = prms[2] - y0_max_sigma
-                    r = np.sqrt((x_diff)**2 + (y_diff)**2)
-                dist_to_centre.append(r)
-                aspect.append(prms[3] / aspect_max_sigma)
-                theta_diff = np.abs(prms[4] - theta_max_sigma) % 360
-                # Because there's redundancy about which way an ellipse 
-                # is aligned, we always take the acute angle
-                if theta_diff > 90:
-                    theta_diff -= 90
-                theta.append(theta_diff)
-
-            return np.hstack((residuals, dist_to_centre, aspect, theta))
+                    max_ind = np.argmax(self.percentiles)
+
+                    residuals = []
+                    dist_to_centre = []
+                    aspect = []
+                    theta = []
+
+                    x0_max_sigma = feats[max_ind][1]
+                    y0_max_sigma = feats[max_ind][2]
+                    aspect_max_sigma = feats[max_ind][3]
+                    theta_max_sigma = feats[max_ind][4]
+
+                    for n in range(len(feats)):
+                        prms = feats[n]
+                        residuals.append(prms[0])
+                        if prms[1] == 0 or prms[2] == 0:
+                            r = 0
+                        else:
+                            x_diff = prms[1] - x0_max_sigma
+                            y_diff = prms[2] - y0_max_sigma
+                            r = np.sqrt((x_diff)**2 + (y_diff)**2)
+                        dist_to_centre.append(r)
+                        aspect.append(prms[3] / aspect_max_sigma)
+                        theta_diff = np.abs(prms[4] - theta_max_sigma) % 360
+                        # Because there's redundancy about which way an ellipse 
+                        # is aligned, we always take the acute angle
+                        if theta_diff > 90:
+                            theta_diff -= 90
+                        theta.append(theta_diff)
+                        features = np.hstack((residuals, dist_to_centre, aspect, theta))
+
+                    if len(lst) == len(percentiles):
```

Comment:
This stops the loop when all of the contours are found  [ len(percentiles) ]. Without this, the loop keeps running for all i in the range and produces unnecessary computations. The alternative for this is to use a fixed up-scale value, but that defeats the purpose of having a dynamic up-scaling aspect. This is basically just a check that when all of the required contours are found it stops and moves to the next source instead of up-scaling further.